src/librustc_mir/interpret/intern.rs

```diff
@@ -224,69 +262,44 @@ for
     }
 }
 
-/// Figure out the mutability of the allocation.
-/// Mutable if it has interior mutability *anywhere* in the type.
-fn intern_mutability<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    param_env: ParamEnv<'tcx>,
-    ty: Ty<'tcx>,
-    span: Span,
-    mutability: Mutability,
-) -> Mutability {
-    let has_interior_mutability = !ty.is_freeze(tcx, param_env, span);
-    if has_interior_mutability {
-        Mutability::Mutable
-    } else {
-        mutability
-    }
-}
-
 pub fn intern_const_alloc_recursive(
     ecx: &mut CompileTimeEvalContext<'mir, 'tcx>,
     def_id: DefId,
     ret: MPlaceTy<'tcx>,
-    // FIXME(oli-obk): can we scrap the param env? I think we can, the final value of a const eval
-    // must always be monomorphic, right?
-    param_env: ty::ParamEnv<'tcx>,
 ) -> InterpResult<'tcx> {
     let tcx = ecx.tcx;
     // this `mutability` is the mutability of the place, ignoring the type
-    let (mutability, base_intern_mode) = match tcx.static_mutability(def_id) {
+    let (base_mutability, base_intern_mode) = match tcx.static_mutability(def_id) {
         Some(hir::Mutability::MutImmutable) => (Mutability::Immutable, InternMode::Static),
         None => (Mutability::Immutable, InternMode::ConstBase),
         // `static mut` doesn't care about interior mutability, it's mutable anyway
         Some(hir::Mutability::MutMutable) => (Mutability::Mutable, InternMode::Static),
     };
```

Comment:
Does this correctly compute `ConstBase` for things like array sizes and non-`Copy` array initializers? What does it do for promoteds?

Reply:
for promoteds we also run into the `None` branch, as the `DefId` is of the function of the promoted, which doesn't have `static_mutability`.

The only way we cannot get `ConstBase` is by the `DefId` pointing to a static. So not sure what happens to `non-Copy` array initializers, as that's promotion happening inside a static

Reply:
> the DefId is of the function of the promoted, which 

But what if we are promoting inside the body of a `static`?

> not sure what happens to non-Copy array initializers, as that's promotion happening inside a static

@eddyb do you know more here?

Reply:
Do we actually create implicit promoteds inside statics? I don't think so. I think we just nuke a bunch of storage statements

Reply:
Hm... seems like that is the case, and also for `const`:
```rust
const fn id<T>(x: T) -> T { x }

const FOO: &'static usize = {
    let v = id(&std::mem::size_of::<i32>());
    v
};
```
[Check the MIR](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e5f4c87bd7b22e1d144f885703f90b87)

But... @eddyb isn't that in contradiction to what you told me about the "enclosing scope" stuff?

Reply:
Maybe I should open a PR about fixing that logic, or an issue at least.
We should only be nuking `StorageDead` for locals that we want to promote the *borrow* of, nothing else.

I guess if I want `[None; 123]` to work for any element type in a constant, we'd have to do *actual* promotion, hmpf, we can't do nothing (since then MIR borrowck will still error like it does today) nor remove the `StorageDead`'s (since that has no effect either).

Reply:
@RalfJung Sorry, everyone is replying and GitHub is not updating, so assume most of my comments are minutes older than GitHub says.

Your use of `id` is entirely unnecessary, introducing an inner `let` drops the `'static` scope anyway, so `&std::mem::size_of::<i32>()` has to be promoted (by removing `StorageDead`).

Note that something like `const X: &T = &expr;` doesn't have `StorageDead` on the local being borrowed *from MIR building*, whereas anything more nested requires promotion *to reach that* (but both use the lack of a `StorageDead` to signal that).

Reply:
> has to be promoted (by removing StorageDead).

Wait, what? I thought "being promoted" means firing up the promotion machinery, where parts of the local MIR are moved into a new body and computed separately and so on. That's what promotion *is*, we even have [a type for that](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/struct.Promoted.html) these days and [a query](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.promoted_mir). Are you saying that use of the term "promoted" is wrong -- or at least, that not all "promoteds"  actually have a corresponding `Promoted`? That would be really bad use of terminology...

It seems that even are lots of discussion, we are still not using the term "promotion" the same way. I'd really appreciate if you could define precisely what you mean by that term, because this is getting frustrating.^^

Reply:
Alternatively I propose we use "promoted" to refer to what is called a `Promoted` in rustc, and then we probably have to call whatever happens in my example above something else. "Lifetime extension" or so.

Reply:
Promotion in consts and statics *is performed* by removing the `StorageDead` - that's what the bitset returned by `mir_const_qualif` is for - since that already has to work.

Maybe it shouldn't, especially since it's not enough for the `[None; 123]` case I mentioned earlier.

It would actually be less code overall to be uniform (but it might be slightly more expensive). I can prepare a PR if you want.

Reply:
> Promotion in consts and statics is performed by removing the StorageDead - that's what the bitset returned by mir_const_qualif is for - since that already has to work.

So I was using the term for the mechanism, and you were using for the effect. That would explain the confusion.
Nuking `StorageDead` actually makes sense to me, and that's what I used to think happens in consts -- but then I talked with you and you taught me about the "enclosing scope" rule and since then I was sure it was *not* done with `StorageDead`.

I think "promotion" makes more sense as the name for the mechanism of extracting part of the MIR out -- and that's what `Promoted` already means in the compiler. So I feel like we should come up with another name for the effect of something having a longer lifetime than expected. But we can also do it the other way around, as long as we are able to clearly distinguish the two.

We are looking for three terms:
* The mechanism that extracts parts of a MIR and moves it out (the kind of thing `promoted_mir` gives access to, from what I understand).
* The specific case of using that mechanism on unsuspecting run-time code, which has the effect of moving it to compile-time, leading to a very restrictive analysis (this is what `promoted.md` in this repo talks about).
* The analysis that determines which things get `'static` lifetime even though they usually wouldn't. The result of this analysis is used to nuke `StorageDead` in `const` (and `static`?), and to extract part of a MIR and cut it out elsewhere.

Finally I feel like we are making progress on this minefield of three different things that are all referred to by the same name.^^

Reply:
So, coming back to the [original question](https://github.com/rust-lang/rust/pull/63955#discussion_r318256605): does this code correctly result in `ConstBase` for all the right things? It seems like the answer is yes right now, but how fragile is that? Does `static_mutability` guarantee to return `None` for "MIR promoteds"?

Reply:
> I guess if I want [None; 123] to work for any element type in a constant, we'd have to do actual promotion, hmpf, we can't do nothing (since then MIR borrowck will still error like it does today) nor remove the StorageDead's (since that has no effect either).

So currently non-`Copy` array initializers don't work in statics/consts?

Reply:
> So currently non-`Copy` array initializers don't work in statics/consts?

Correct, I don't see how they could.

Reply:
[Looks like you are right](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=cbb7b9a3a70fcdc0578109b425b9bbab).

Are there other cases we have to worry about here in terms of classifying CTFE-evaluated "stuff" into statics and consts ("everything but statics")?

Reply:
Kind of related is that consts can't refer to statics

Reply:
How is that related? Even if they would, that memory would already be interned.

Reply:
It was the only thing I could come up with where we decide only on explict static vs not

Reply:
"explict static vs not"?